[GStreamer] registerAppsinkWithWorkaroundsIfNeededCallOnce() should be callable from non-main threads

philn opened the pull request:
#### c760465095dda52270369a1f4f620a253bc320d8
<pre>
[GStreamer] registerAppsinkWithWorkaroundsIfNeededCallOnce() should be callable from non-main threads
<a href="https://bugs.webkit.org/show_bug.cgi?id=260711">https://bugs.webkit.org/show_bug.cgi?id=260711</a>

Reviewed by Alicia Boya Garcia.

Remove un-needed ASSERT, nothing should prevent to call gst_element_register() from a non-main thread.

* Source/WebCore/platform/AudioDecoder.cpp:
(WebCore::AudioDecoder::isCodecSupported):
* Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp:
(WebCore::registerAppsinkWithWorkaroundsIfNeededCallOnce):

Canonical link: <a href="https://commits.webkit.org/267550@main">https://commits.webkit.org/267550@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab46ad29fefe4ce76668675244182dda64372dec

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16943 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17266 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17727 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18727 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15862 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17134 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20498 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17405 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17143 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17495 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14671 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19539 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14738 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15366 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22089 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15722 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15533 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19851 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16141 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13675 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15303 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4057 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19668 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15971 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ntrrgc reviewed (commented):
Seeing this I really need to ask, in what cases is GStreamer being initialized from a non-main thread, and is that something we really want?

Analyzing the code a little bit, this seems to be the series of steps triggering the failed assert:

1. `GStreamerRegistryScanner::singleton()` is called from a non-main thread, which contains this code:

   ```cpp
   GStreamerRegistryScanner& GStreamerRegistryScanner::singleton()
   {
       static NeverDestroyed<GStreamerRegistryScanner> sharedInstance;
       singletonInitialized = true;
       return sharedInstance;
   }
   ```

   `sharedInstance` is initialized the first time the function is called. Fortunately the compiler actually places thread-safe guards here.

2. This is the body of the constructor:

   ```cpp
   GStreamerRegistryScanner::GStreamerRegistryScanner(bool isMediaSource)
       : m_isMediaSource(isMediaSource)
   {
       if (isInWebProcess())
           ensureGStreamerInitialized();
       else {
           // This is still needed, mostly because of the webkit_web_view_can_show_mime_type() public API (so
           // running from UIProcess).
           gst_init(nullptr, nullptr);
       }
   
       GST_DEBUG_CATEGORY_INIT(webkit_media_gst_registry_scanner_debug, "webkitregistryscanner", 0, "WebKit GStreamer registry scanner");
   
       refresh();
   }
   ```

3. Finally, `ensureGStreamerInitialized` uses a `std::call_once()` where `gst_init_check()` is called, and after, `registerAppsinkWithWorkaroundsIfNeeded()`.

So far I haven't seen anything causing a potential race with the GStreamer initialization (both GStreamer's and ours) in a random secondary thread, but I wouldn't be surprised if there was. Allowing the GStreamer backend to initialize in any thread sounds more risk than it's worth.

ntrrgc reviewed (approved):
I'll aprove the change but recommend a comment be placed in our GStreamer initialization code making it explicit it may run in any thread.

philn committed:
[GStreamer] registerAppsinkWithWorkaroundsIfNeededCallOnce() should be callable from non-main threads
https://bugs.webkit.org/show_bug.cgi?id=260711

Reviewed by Alicia Boya Garcia.

Remove un-needed ASSERT, nothing should prevent to call gst_element_register() from a non-main thread.

* Source/WebCore/platform/AudioDecoder.cpp:
(WebCore::AudioDecoder::isCodecSupported):
* Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp:
(WebCore::registerAppsinkWithWorkaroundsIfNeededCallOnce):

Canonical link: https://commits.webkit.org/267550@main
